### PR TITLE
fix: more consistent disposal of resources at the end of the test

### PR DESCRIPTION
**Pull request by keturn**
A few things here:

We were putting ModuleTestingHelper instances _in_ to this map and never taking them out: https://github.com/Terasology/ModuleTestingEnvironment/blob/38100cf07da848e13103e0b599cb69b881f5a12b/src/main/java/org/terasology/moduletestingenvironment/MTEExtension.java#L54

There were scenarios where we'd be trying to use old instances _after_ their `tearDown`. And awkward situations like this where it wouldn't consistently be ready before `postProcessTestInstance`: https://github.com/Terasology/ModuleTestingEnvironment/blob/38100cf07da848e13103e0b599cb69b881f5a12b/src/main/java/org/terasology/moduletestingenvironment/IsolatedMTEExtension.java#L38-L41

This PR addresses these problems with managing the lifecycle of the ModuleTestingHelper instance by using the store mechanism on jupiter's ExtensionContext. See [JUnit 5 User Guide: Keeping State in Extensions](https://junit.org/junit5/docs/current/user-guide/#extensions-keeping-state).

Required to fix tests in https://github.com/Terasology/DynamicCities/pull/94  
and likely useful for Pathfinding tests as well.

## To Do
- [x] write test for class lifecycle
- [x] javadoc

### CODE REVIEW

File: src/main/java/org/terasology/moduletestingenvironment/IsolatedMTEExtension.java [hunk 1]
```diff
@@ -17,27 +12,9 @@
...
-        super.afterAll(context);
-    }
-
-    @Override
-    public void postProcessTestInstance(Object testInstance, ExtensionContext extensionContext) throws Exception {
-        // beforeEach would be run after postProcess so postProcess would NPE, so we initialize the MTE here beforehand
-        super.beforeAll(extensionContext);
-        super.postProcessTestInstance(testInstance, extensionContext);
+    protected ExtensionContext.Namespace getNamespace(ExtensionContext context) {
+        return ExtensionContext.Namespace.create(MTEExtension.class, context.getTestMethod());
```
Comment (keturn) [inline src/main/java/org/terasology/moduletestingenvironment/IsolatedMTEExtension.java]: TODO: comment

File: src/main/java/org/terasology/moduletestingenvironment/MTEExtension.java [hunk 2]
```diff
@@ -140,6 +105,44 @@ public void postProcessTestInstance(Object testInstance, ExtensionContext extens
         }
     }
 
+    private static ModuleTestingHelper setupHelper(ExtensionContext context) {
```
Comment (keturn) [inline src/main/java/org/terasology/moduletestingenvironment/MTEExtension.java]: TODO: doc

File: src/main/java/org/terasology/moduletestingenvironment/MTEExtension.java [hunk 3]
```diff
@@ -140,6 +105,44 @@ public void postProcessTestInstance(Object testInstance, ExtensionContext extens
...
+            helperContext.setup();
+        } catch (RuntimeException e) {
+            throw e;
+        } catch (Exception e) {
+            throw new RuntimeException(e);
+        }
+        return helperContext;
+    }
+
+    private ModuleTestingHelper getHelper(ExtensionContext context) {
```
Comment (keturn) [inline src/main/java/org/terasology/moduletestingenvironment/MTEExtension.java]: TOOD: doc

File: src/main/java/org/terasology/moduletestingenvironment/MTEExtension.java [hunk 4]
```diff
@@ -140,6 +105,44 @@ public void postProcessTestInstance(Object testInstance, ExtensionContext extens
...
+        ExtensionContext.Store store = context.getStore(getNamespace(context));
+        HelperCleaner autoCleaner = store.getOrComputeIfAbsent(
+                HelperCleaner.class, k -> new HelperCleaner(setupHelper(context)),
+                HelperCleaner.class);
+        return autoCleaner.helper;
+    }
+
+    protected ExtensionContext.Namespace getNamespace(ExtensionContext context) {
```
Comment (keturn) [inline src/main/java/org/terasology/moduletestingenvironment/MTEExtension.java]: TODO: doc

File: src/main/java/org/terasology/moduletestingenvironment/MTEExtension.java [hunk 5]
```diff
@@ -140,6 +105,44 @@ public void postProcessTestInstance(Object testInstance, ExtensionContext extens
...
+        return ExtensionContext.Namespace.create(MTEExtension.class, getTopTestClass(context));
+    }
+
+    protected static Class<?> getTopTestClass(ExtensionContext context) {
```
Comment (keturn) [inline src/main/java/org/terasology/moduletestingenvironment/MTEExtension.java]: TODO: doc

File: src/main/java/org/terasology/moduletestingenvironment/MTEExtension.java [hunk 6]
```diff
@@ -174,4 +177,18 @@ void setupLogging() {
             StatusPrinter.printInCaseOfErrorsOrWarnings(context);
         }
     }
+
+    static class HelperCleaner implements ExtensionContext.Store.CloseableResource {
```
Comment (keturn) [inline src/main/java/org/terasology/moduletestingenvironment/MTEExtension.java:185]: TODO: doc

**Review by skaldarnar — COMMENTED** [review]
I don't get the details, but it looks cleaner overall 👍  

Wrapping the `helper` into a closeable resource looks like a good move.

**Review by keturn — COMMENTED** [review]
While I was documenting the methods, I found another few revisions to make, so this has changed a bit from the last time you last looked, @skaldarnar.

There are some terminology issues that could use straightening out. Part of it is the vagueness of "ModuleTestingHelper." There are places where something like "engine" or "host" might be clearer, but it claims to support _multiple_ engines (host + client(s)), so calling it "engine" doesn't fit.

Then there's the terminology around how widely an instance of that Helper is shared. We've used the word "isolation" to describe that, and it's now implemented with Namespaces, so I thought of it as "scope."

But JUnit describes a similar thing as "lifecycle," i.e. something defining how long the resource lives.

Renames are easy, as these names are rarely (if ever) used by outside code, so we don't have to figure all the naming out now. But do chime in if you have any ideas.

**Comment by keturn** [comment]
Oh, and the name "Cleaner" is a bit less accurate now that I decided that's a good place to throw some additional code that serves as an adapter between MTEExtension and a Helper instance.

I wouldn't spend too long on naming that one; I expect the methods I stuffed in there will change quite a bit after https://github.com/Terasology/DynamicCities/pull/94 drops the last reference to the old class and we can easily split it up.

**Review by skaldarnar — APPROVED** [review]

**Comment by keturn** [comment]
[MTE tests: MTE](https://trello.com/c/P7W00HX2/177-mte-tests-mte)

**Pull request merged**